### PR TITLE
Add a condition for parsing zero from string when not denominated.

### PR DESCRIPTION
**Pull request by yancyribbens**
closes https://github.com/rust-bitcoin/rust-bitcoin/issues/3307

### CODE REVIEW

File: units/src/amount.rs [hunk 1]
```diff
@@ -1158,7 +1158,13 @@ impl ops::DivAssign<u64> for Amount {
 impl FromStr for Amount {
     type Err = ParseError;
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> { Amount::from_str_with_denomination(s) }
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if !s.chars().any(char::is_alphabetic) && s.parse::<f64>() == Ok(0.0) && !s.starts_with('-') {
```
Comment (tcharding) [inline units/src/amount.rs:1162]: I can not easily reason about the statement `!s.chars().any(char::is_alphabetic)`

The unit tests seem to implement what was discussed in the issue and the following code passes the tests and is more simple

```rust
        // `Amount` is an unsigned type so we do not allow `-`.
        if s.parse::<f64>() == Ok(0.0) && !s.starts_with('-') {
            Ok(Amount::ZERO)
        } else {
            Amount::from_str_with_denomination(s)
        }
```

**Review by Kixunil — CHANGES_REQUESTED** [review]
Please do it as I recommended in the issue (try to parse it with denomination and if it fails also try the usual parsing function and check for zero). The way you wrote it makes the code possibly inconsistent and I suspect it'd break for inputs like `0.00000000000000001`.

File: units/src/amount.rs (hunk 1 above)
Comment (Kixunil) [reply units/src/amount.rs:1162]: Yeah, that condition is useless. But we should also have a test that the smallest non-zero value (as restricted by max len) does not parse.

Comment (yancyribbens) [reply units/src/amount.rs]: @tcharding thanks.  The condition was to guard against anything that did have a denomination from being parsed as float, although it's not needed anymore.

**Comment by yancyribbens** [comment]
@Kixunil I've updated it as you suggested.  It's a bit more verbose but I understand wanting consistency.  I'm kinda surprised not just using standard float parsing under the hood, although I imagine there is a reason..

**Comment by yancyribbens** [comment]
It looks like alloc is now required unless I'm missing something.

**Comment by Kixunil** [comment]
> I'm kinda surprised not just using standard float parsing under the hood, although I imagine there is a reason..

The reason is floats are absolutely wrong for financial applications because they are not precise enough. There's a ton about this on the Internet, I'm sure you'll find more information easily if you need it.

`alloc` shouldn't be needed but it might if we have input string that is not `InputString`. I'll look into it later, I can only do surface-level stuff right now.

**Comment by yancyribbens** [comment]
> The reason is floats are absolutely wrong for financial applications because they are not precise enough. There's a ton about this on the Internet, I'm sure you'll find more information easily if you need it.

Hmm that's interesting.  I read that financial applications usually use decimal notation with fixed precision vs a float where the precision "floats".  I take it that means the rust library `rust_decimal` would be comparable to Amounts numeric representation.

> alloc shouldn't be needed but it might if we have input string that is not InputString. I'll look into it later, I can only do surface-level stuff right now.

The source string "0" has " sat" concatenated to the end so that it doesn't fail to parse.  somehow there needs to be an allocation so that the source string can resize to hold the denomination which is why I think it needs alloc, unless Im mistaken.

File: units/src/amount.rs [hunk 2]
```diff
@@ -1155,10 +1165,27 @@ impl ops::DivAssign<u64> for Amount {
     fn div_assign(&mut self, rhs: u64) { *self = *self / rhs }
 }
 
+#[cfg(feature = "alloc")]
 impl FromStr for Amount {
     type Err = ParseError;
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> { Amount::from_str_with_denomination(s) }
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let a = Amount::from_str_with_denomination(s);
```
Comment (Kixunil) [inline units/src/amount.rs:1173]: Please name the variable `result`.

File: units/src/amount.rs [hunk 3]
```diff
@@ -1155,10 +1165,27 @@ impl ops::DivAssign<u64> for Amount {
...
+
+        match a {
+            Err(ParseError::MissingDenomination(_)) => {
+                let d = s.to_string() + " sat";
+                let w = Amount::from_str_with_denomination(&d);
```
Comment (Kixunil) [inline units/src/amount.rs:1178]: Don't do these string operations, call `Amount::from_str_in(Denomination::Satoshi)` instead.

File: units/src/amount.rs [hunk 4]
```diff
@@ -1574,10 +1601,27 @@ impl ops::Neg for SignedAmount {
...
     type Err = ParseError;
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> { SignedAmount::from_str_with_denomination(s) }
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let a = SignedAmount::from_str_with_denomination(s);
+
+        match a {
+            Err(ParseError::MissingDenomination(_)) => {
+                let d = s.to_string() + " sat";
+                let w = SignedAmount::from_str_with_denomination(&d);
```
Comment (Kixunil) [inline units/src/amount.rs:1614]: Same issues as above.

File: units/src/amount.rs (hunk 3 above)
Comment (yancyribbens) [reply units/src/amount.rs]: Ah cool, that's how to avoid needing alloc.

File: units/src/amount.rs (hunk 2 above)
Comment (yancyribbens) [reply units/src/amount.rs]: result seems kinda verbose, why not just r?  Anyway not worth nit picking about so I've named it result.

File: units/src/amount.rs (hunk 4 above)
Comment (yancyribbens) [reply units/src/amount.rs]: complete

**Comment by yancyribbens** [comment]
> Also this needs documenting how FromStr differs from from_str_with_denomination. I think the difference makes sense. I've briefly considered deprecating from_str_with_denomination but it looks like it might be useful to keep it.

I would have thought there be a compile error about `from_str` not being documented.  Anyway, now the difference I see is just that `from_str` doesn't imply denomination is required, however the only case where that makes sense is with zero.  Also its counter intuitive the way the naming is done using preposition like `with` to mean that the denomination is provided in the string and not as formal argument.  For example, `from_str_in` takes denom as an argument, so I would have thought this would be named `.._with_denom`.  I think `from_str_with_denomination` is more precise as `from_str_include_denomination`.

**Comment by yancyribbens** [comment]
> /// If you want to parse only the amount without the denomination, use [`Self::from_str_in`].

On topic, `f you want to parse only the amount given a denomination parameter, ` or something like that is better imo

File: units/src/amount.rs [hunk 5]
```diff
@@ -2075,6 +2105,25 @@ mod tests {
...
+        assert!(Amount::from_str("-00.0").is_err());
+
+        let _a = SignedAmount::from_str("-0").unwrap();
+        let _a = SignedAmount::from_str("-0.0").unwrap();
+        let _a = SignedAmount::from_str("-00.0").unwrap();
+
+        let _a = SignedAmount::from_str("0").unwrap();
+        let _a = SignedAmount::from_str("0.0").unwrap();
+        let _a = SignedAmount::from_str("00.0").unwrap();
+    }
```
Comment (Kixunil) [inline units/src/amount.rs:2125]: A test confirming that `0.00` parses would be nice too.

File: units/src/amount.rs [hunk 6]
```diff
@@ -1158,7 +1158,22 @@ impl ops::DivAssign<u64> for Amount {
...
+            Err(ParseError::MissingDenomination(_)) => {
+                let d = Amount::from_str_in(s, Denomination::Satoshi);
+
+                if d == Ok(Amount::ZERO) {
+                    Ok(Amount::ZERO)
+                } else {
+                   result
+                }
+            },
+            _ => result
```
Comment (Kixunil) [inline units/src/amount.rs:1174]: I'm amazed the compiler doesn't complain about this being moved.

**Review by Kixunil — APPROVED** [review]
ACK 894f82e7cc9eb459a297d43e82734621e0824610

File: units/src/amount.rs (hunk 6 above)
Comment (tcharding) [reply units/src/amount.rs:1174]: This is pretty clever of the borrow checker. I bet there is some long PR/issue somewhere with people arguing about how the borrow checker should be able to work out that the error (non-`Copy`) is never used other than in paths that just return it.

**Review by tcharding — APPROVED** [review]
ACK 894f82e7cc9eb459a297d43e82734621e0824610

File: units/src/amount.rs (hunk 6 above)
Comment (yancyribbens) [reply units/src/amount.rs:1174]: A &str slice is a read-only copy and doesn't actually own the data like a heap pointer would.  I think these pointers are actually just to read only parts of the binary itself.  Could be wrong or just making stuff up though.

File: units/src/amount.rs (hunk 5 above)
Comment (yancyribbens) [reply units/src/amount.rs:2125]: can just do in another comment since this one is already acked.  This is also something that could just be part of a proptest given arbitrary dot placement which is probably more effective.

File: units/src/amount.rs (hunk 6 above)
Comment (yancyribbens) [reply units/src/amount.rs:1174]: A wait this returns Amount not &str

File: units/src/amount.rs (hunk 5 above)
Comment (Kixunil) [reply units/src/amount.rs:2125]: Good idea turning it to proptest, but I'd like to see also various lengths not just dot placement being changed.

File: units/src/amount.rs (hunk 6 above)
Comment (Kixunil) [reply units/src/amount.rs:1174]: It returns `Result` which contains `Error`, which is not `Copy` IIRC.

I think it has something to do with the `_` pattern.

**Review by apoelstra — COMMENTED** [review]
ACK 894f82e7cc9eb459a297d43e82734621e0824610; successfully ran local tests.

File: units/src/amount.rs (hunk 6 above)
Comment (yancyribbens) [reply units/src/amount.rs:1174]: Err yes, it returns result.  I had meant to return to this conversation yesterday but got distracted.  Anyway, Result implements Copy: https://doc.rust-lang.org/1.81.0/src/core/result.rs.html#523.  I'm not sure why `_` would matter in this case.
